### ISSUE TITLE
HRPO gates a vocabulary-projected soft token, not the previous hidden state

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#69]
## Reviewed source

Default-branch commit: [11b700ba0c9cc03b0e1a54db15fe06d15fad8e51](https://github.com/microprediction/latentreasoning/commit/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51)

Affected text:

- [training.html lines 352–370](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L352-L370)
- [bibliography.html lines 1129–1132](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/bibliography.html#L1129-L1132)

Both summaries say that HRPO “mixes” or “gates the previous hidden state into the sampled token's embedding.” The training page then says the hidden state “rides along” as a continuous side channel.

I searched open and closed issues for HRPO, hidden-state gating, vocabulary projection, and soft tokens; I found no existing issue covering this distinction.

## The exact HRPO transition does not feed back the hidden state

In current [Yue et al., arXiv:2505.18454v2, §3.1, equations 2–4](https://arxiv.org/html/2505.18454#S3.SS1), the final hidden state first passes through the ordinary language head:

~~~text
p_(t+1) = softmax(Head(hat_h_t) / temperature)
~~~

HRPO then constructs

~~~text
h_(t+1) = E^T p_(t+1) / ||p_(t+1)||
~~~

and gates that vector together with the sampled token embedding. Thus the continuous branch is a normalized probability-weighted mixture of vocabulary embeddings—a **soft token**—not the raw hidden state itself.

The released implementation confirms the equation. At commit [8b49045](https://github.com/Yueeeeeeee/HRPO/commit/8b49045004fa34c4bfd12922fa3c792dac5479d1), generation computes its variable named last_thinking_states as an einsum of the next-token probabilities with the input-embedding matrix and normalizes it by the probability-vector norm:

- [transformers/generation/utils.py lines 3367–3372](https://github.com/Yueeeeeeee/HRPO/blob/8b49045004fa34c4bfd12922fa3c792dac5479d1/transformers/generation/utils.py#L3367-L3372)

That soft-token vector is what the residual gate receives:

- [unsloth/models/llama.py lines 940–949](https://github.com/Yueeeeeeee/HRPO/blob/8b49045004fa34c4bfd12922fa3c792dac5479d1/unsloth/models/llama.py#L940-L949)

The source paper itself sometimes uses “hidden state” loosely in prose, but its displayed method and code are unambiguous.

## Why the distinction matters

The route is

~~~text
hidden state -> vocabulary logits -> vocabulary distribution -> soft token -> gate
~~~

so HRPO discards whatever information is not retained by that vocabulary-mediated projection before the next step. The continuous side channel carries the full next-token distribution rather than only the sampled token, but it is not an unconstrained hidden-state recurrence like Coconut.

The current page's conclusion that HRPO is a hybrid rather than a fully latent method is therefore directionally right, but for a stronger reason than stated: **both branches are vocabulary-mediated**. One branch samples a hard token; the other feeds back a soft mixture over token embeddings.

That also makes HRPO closer architecturally to Soft Thinking / mixture-of-token methods than the “previous hidden state rides along” wording implies.

## Consequence

The site's page is specifically classifying where stochasticity and latent state live. Calling the feedback branch a prior hidden state overstates the amount and kind of continuous information that survives between steps, and obscures the method's vocabulary bottleneck.

The reported benchmark comparisons and the claim that token sampling supplies stochasticity need not change.

## Concrete correction

For example:

> HRPO samples a token normally, but it also maps the previous final hidden state through the language head to a full vocabulary distribution and converts that distribution into a normalized soft token. A learnable gate mixes this soft token with the sampled token embedding. Token sampling supplies the stochastic action for REINFORCE; the continuous branch carries more information than the sampled token but remains vocabulary-mediated rather than feeding back the raw hidden state.

The bibliography's final sentence could then say that the loop closes through the vocabulary on **both** the hard and soft branches.

**Comment by microprediction** [comment]
Correct, fixed in 1185867, on both the training page and the bibliography.

Verified against the paper and the released code. Equation 3 is h_{t+1} = W_e^T p_{t+1} / ||p_{t+1}|| with p_{t+1} = softmax(Head(h_t)/tau), and the paper gives the reason: "hidden states often lie outside the model's token embedding manifold, which degrades generation quality when fed directly." The code confirms it, computing the probability-weighted embedding average with einsum and normalising before the gate sees it.

So the gated quantity is a vocabulary-projected soft token, not the hidden state. The paper's abstract does say "integrates prior hidden states", which is where the loose reading came from, but the method section is unambiguous and the page should follow the method section.
